src/execution/execute.ts

```diff
@@ -866,29 +906,62 @@ function ensureValidRuntimeType(
 
   if (typeof runtimeTypeName !== 'string') {
     throw new GraphQLError(
-      `Abstract type "${returnType.name}" must resolve to an Object type at runtime for field "${info.parentType.name}.${info.fieldName}" with ` +
-        `value ${inspect(result)}, received "${inspect(runtimeTypeName)}".`,
+      `Abstract type resolution for "${returnType.name}" for field "${info.parentType.name}.${info.fieldName}" ` +
+        `with value ${inspect(result)} failed. ` +
+        `Encountered abstract type "${abstractType.name}" must resolve to an Object or Interface type at runtime, ` +
+        `received "${inspect(runtimeTypeName)}".`,
+    );
+  }
+
+  if (encounteredTypeNames.has(runtimeTypeName)) {
+    throw new GraphQLError(
+      `Abstract type resolution for "${returnType.name}" for field "${info.parentType.name}.${info.fieldName}" failed. ` +
+        `Encountered abstract type "${abstractType.name}" resolved to "${runtimeTypeName}", causing a cycle.`,
     );
   }
+  encounteredTypeNames.add(runtimeTypeName);
 
   const runtimeType = exeContext.schema.getType(runtimeTypeName);
   if (runtimeType == null) {
     throw new GraphQLError(
-      `Abstract type "${returnType.name}" was resolved to a type "${runtimeTypeName}" that does not exist inside the schema.`,
+      `Abstract type resolution for "${returnType.name}" for field "${info.parentType.name}.${info.fieldName}" failed. ` +
+        `Encountered abstract type "${abstractType.name}" was resolved to a type "${runtimeTypeName}" that does not exist inside the schema.`,
       { nodes: fieldNodes },
     );
   }
 
+  if (isInterfaceType(runtimeType)) {
+    if (!exeContext.schema.isSubType(returnType, runtimeType)) {
+      throw new GraphQLError(
+        `Abstract type resolution for "${returnType.name}" for field "${info.parentType.name}.${info.fieldName}" failed. ` +
+          `Interface type "${runtimeType.name}" is not a subtype of encountered interface type "${returnType.name}".`,
+        { nodes: fieldNodes },
+      );
+    }
+
+    return resolveType(
+      exeContext,
+      returnType,
+      runtimeType,
+      fieldNodes,
+      info,
+      result,
+      encounteredTypeNames,
+    );
+  }
+
   if (!isObjectType(runtimeType)) {
     throw new GraphQLError(
-      `Abstract type "${returnType.name}" was resolved to a non-object type "${runtimeTypeName}".`,
+      `Abstract type resolution for "${returnType.name}" for field "${info.parentType.name}.${info.fieldName}" failed. ` +
+        `Encountered abstract type "${abstractType.name}" was resolved to a non-object type "${runtimeTypeName}".`,
       { nodes: fieldNodes },
     );
   }
 
   if (!exeContext.schema.isSubType(returnType, runtimeType)) {
     throw new GraphQLError(
-      `Runtime Object type "${runtimeType.name}" is not a possible type for "${returnType.name}".`,
+      `Abstract type resolution for "${returnType.name}" for field "${info.parentType.name}.${info.fieldName}" failed. ` +
```

Comment:
You still need to check if the intermediate interface is a sub-type of `returnType` and currently you bypassing this check.
Also please add a test for it.

Reply:
There is a test a few lines above for the intermediate interface.

How is the test being bypassed?

Reply:
ah, you meant that i am checking against `returnType` but not the current abstract type. fixed, new test added.

We should get an error if we go from Named => Pet => Person, even though Person is Named, it's not a Pet.

df8f15c7

Reply:
7ceffbe8

Reply:
7ceffbe8

Reply:
1e2988e